src/snowflake/connector/cache.py

```diff
@@ -260,3 +306,185 @@ def _expiration(self, k: K) -> None:
         called from contexts where the lock is already held.
         """
         self.telemetry["expiration"] += 1
+
+
+class SFDictFileCache(SFDictCache):
+    def __init__(
+        self,
+        file_path: str | dict[str, str],
+        entry_lifetime: int = constants.DAY_IN_SECONDS,
+        file_timeout: int = 0,
+    ) -> None:
+        """Inits an SFDictFileCache with path, lifetime and .
+
+        File path can be a dictionary that contains different paths for different OSes,
+        possible keys are: 'darwin', 'linux' and 'windows'. If a current platform
+        cannot be determined, or is not in the dictionary we'll use the first value.
+        """
+        super().__init__(
+            entry_lifetime=entry_lifetime,
+        )
+        if isinstance(file_path, str):
+            self.file_path = os.path.expanduser(file_path)
+        else:
+            current_platform = platform.system().lower()
+            if current_platform is None or current_platform not in file_path:
+                self.file_path = next(iter(file_path.values()))
+            else:
+                self.file_path = os.path.expanduser(file_path[current_platform])
+        self.file_timeout = file_timeout
+        self._file_lock_path = f"{self.file_path}.lock"
+        self._file_lock = FileLock(self._file_lock_path, timeout=self.file_timeout)
+        self.last_loaded: datetime.datetime | None = None
+        if os.path.exists(self.file_path):
+            self._load()
+
+    def __getitem__(self, k: K) -> V:
+        """Returns an element if it hasn't expired yet in a thread-safe way."""
+        self._lock.acquire()
+        # TODO: This variable could be replaced by a wrapper class that keeps track
+        #  of whether the lock is locked, but unless this function gets extended I
+        #  feel like it's an overkill. Make sure to change the bool right after
+        #  self._lock.acquire() and self._lock.release().
+        currently_holding = True
+        try:
+            try:
+                t, v = self._cache[k]
+            except KeyError:
+                # Load if we should and try to see if we can turn the miss into a hit
+                self._lock.release()
```

Comment:
is the re-release/re-acquire necessary here? I don't see `_load_if_should` use `self._lock`

Reply:
`self._lock` is for thread safety, it should not be held while doing file operations.
Holding it while trying to acquire `self._file_lock` could lead to a deadlock.

Reply:
I don't see `_load_if_should` tries to acquire `self._file_lock` (as you explained load doesn't require file lock) so deadlock shouldn't be an issue?

also I think actually locking the whole scope helps avoid reloading the same version of file multiple times.

let's say if we have 2 threads reached here at the same time trying to acquire the lock here, in the current implementation, the file might be loaded twice as _should_load is not protected by lock. As contrast, if we don't release, then the second thread will block (which could be a bad thing I admit), and when it acquire the lock it will  find the file cache is already reloaded to the latest saving one time of read.

Reply:
I don't see `_load_if_should` tries to acquire `self._file_lock` (as you explained load doesn't require file lock) so deadlock shouldn't be an issue?

also I think actually locking the whole scope helps avoid reloading the same version of file multiple times.

let's say if we have 2 threads reached here at the same time trying to acquire the lock here, in the current implementation, the file might be loaded twice as _should_load is not protected by lock. As contrast, if we don't release, then the second thread will block (which could be a bad thing I admit), and when it acquire the lock it will  find the file cache is already reloaded to the latest saving one time of read.

Reply:
You're right, nothing tries to acquire both locks in the current implementation. This was also done to help future developers working with this code.
My thinking was also to minimize the time we hold `self._lock` for, which means that we cannot perform file operations if `self._lock` is held as you point out exactly.

I agree, in the case where multiple threads (in one process) are missing some keys that are all now saved on disk, but be aware that this will affect performance in other cases, like a thread performing a lookup that results in a hit. That is if memory has: `cache_in_memory={"a": 1, "b": 2}, cache_on_disk={"a": 1, "b": 2, "c": 3}` when thread 1 tries to lookup `d` thread 2 won't be able to resolve `cache["a"]` until thread 1 is done loading.